### ISSUE TITLE
Highlight notification in current buffer

### ISSUE THREAD

**Issue by gene1wood** [glowing-bear/glowing-bear#976]
(If this is solved via weechat and not glowing bear, apologies for putting this question in the wrong place, it's my best guess that it's in glowing bear)

When I have glowing bear open in a Firefox tab on my Ubuntu 16.04 desktop, if someone highlights me in a buffer other than the one I'm currently looking at, I get a [web API notification](https://developer.mozilla.org/en-US/docs/Web/API/notification). If someone highlights me in the current buffer open in glowing bear I get no web API notification. This makes sense as I'm already looking at the buffer and would see the highlighted message via the glowing bear highlight colors.

If however, I change windows to another application (not my browser), and in the background my browser with the glowing bear tab is on a buffer for channel #example and someone highlights me in #example, I don't get a web API notification and don't know that I've been highlighted.

Same is true if I'm using the browser but I'm looking at a tab other than glowing bear.

Is this intentional? Do I have something misconfigured in glowing bear (or weechat) such that I don't get web api notifications of highlights in the current buffer? Is this something I can enable?

I'd happily trade the downside of getting web API notifications of highlights in the current buffer when I'm actively using the buffer to get the upside of getting notifications for highlights in the current buffer when I'm in another tab or using another application.

**Comment by lorenzhs** [comment]
It's not intentional, I believe it is (or was?) a limitation of the [Page Visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API). Basically the browser won't tell us. The relevant variables in [glowingbear.js](../tree/master/js/glowingbear.js) are `documentHidden` and `documentVisibilityChange`, and the `isWindowFocused` function.

Maybe there's a better way?

**Comment by gene1wood** [comment]
It looks like it is possible to send a web API notification from an active tab (I tested with the mdn example code [here](http://mdn.github.io/to-do-notifications/) by creating a task, leaving the tab active and getting a notification).

@lorenzhs It looks like the [Page Visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) lets you know if the tab is currently active, but what I'm wondering is if it's possible to tell Glowing Bear to *ignore* whether the page is visible or not and instead *always* create notifications when highlight terms show up (to address the case where glowing bear is the active tab but the browser itself is not visible)

**Comment by lorenzhs** [comment]
Sending notifications when GB is the active tab would be super annoying if you're using private messages, I'm not sure that's the best option... Do you know how other browser-based applications handle this?

**Comment by gene1wood** [comment]
> Sending notifications when GB is the active tab would be super annoying if you're using private messages, I'm not sure that's the best option

Indeed, the best solution would be to detect if the browser containing the tab is the active window. If the browser is not the active window, then regardless of whether the page visibility API says the tab is the active tab, generate a web api notification.

For myself, I would be willing to switch on a preference that had the downside of a notification for every single private message in order to get the upside of notifications for highlights on the active tab when the window is not active (but that's just me).

> Do you know how other browser-based applications handle this?

Yes, I've tested and Slack handles this correctly. In Slack
* If the browser is active and the slack tab is active, an @mention doesn't trigger a notification
* If the browser is active but the slack tab is not active, a notification is triggered
* If the browser is not active but the slack tab is active, a notification is triggered
* If the browser is not active and the slack tab is not active, a notification is triggered

Also here's a [humorous diagram](https://pbs.twimg.com/media/C6ROe0mU0AEmpzz.jpg:large) of some of the logic that goes into slack notifications. It doesn't cover these active / not active cases though.

**Comment by oakkitten** [comment]
what about [`document.hasFocus()`](https://jsfiddle.net/18hkedkf/)?

**Comment by lorenzhs** [comment]
It seems that despite being intended to replace `onFocus` / `onBlur`, the page visibility API really isn't a great replacement for our purpose. Maybe we should go back to checking `hasFocus` in `isWindowFocused`...
